resources/aws_ec2.go

```diff
@@ -0,0 +1,535 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	errwrap "github.com/pkg/errors"
+)
+
+func init() {
+	RegisterResource("aws:ec2", func() Res { return &AwsEc2Res{} })
+}
+// AwsPrefix is a const which gets prepended onto the instance name.
+const AwsPrefix = "mgmt:"
+
+// AwsRegions is a list of all AWS regions generated using ec2.DescribeRegions.
+var AwsRegions = []string{
+	"ap-south-1",
+	"eu-west-2",
+	"eu-west-1",
+	"ap-northeast-2",
+	"ap-northeast-1",
+	"sa-east-1",
+	"ca-central-1",
+	"ap-southeast-1",
+	"ap-southeast-2",
+	"eu-central-1",
+	"us-east-1",
+	"us-east-2",
+	"us-west-1",
+	"us-west-2",
+}
+
+// AwsEc2Res is an AWS EC2 resource.
+type AwsEc2Res struct {
+	BaseRes `yaml:",inline"`
+	State   string `yaml:"state"` // state: running, stopped, terminated
+	Region  string `yaml:"region"`
+	Type    string `yaml:"type"` // the ec2 instance type, ie. t2.micro
+	ImageID string `yaml:"imageid"`
+
+	client *ec2.EC2 // client session for AWS API calls
+	wg     *sync.WaitGroup
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *AwsEc2Res) Default() Res {
+	return &AwsEc2Res{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *AwsEc2Res) Validate() error {
+	if obj.State != "running" && obj.State != "stopped" && obj.State != "terminated" {
+		return fmt.Errorf("state must be 'running', 'stopped' or 'terminated'")
+	}
+
+	// compare obj.Region to the list of available AWS endpoints.
+	validRegion := false
+	for _, region := range AwsRegions {
+		if obj.Region == region {
+			validRegion = true
+		}
+	}
+	if !validRegion {
+		return fmt.Errorf("region must be a valid AWS endpoint")
+	}
+
+	// check the instance type
+	// there is currently no api call to enumerate available instance types
+	if obj.Type == "" {
+		return fmt.Errorf("no instance type specified")
+	}
+
+	// check imageId against a list of available images
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	})
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating session")
+	}
+	client := ec2.New(sess)
+
+	imagesInput := &ec2.DescribeImagesInput{}
+	images, err := client.DescribeImages(imagesInput)
+	if err != nil {
+		return errwrap.Wrapf(err, "error describing images")
+	}
+	validImage := false
+	for _, image := range images.Images {
+		if obj.ImageID == *image.ImageId {
+			validImage = true
+		}
+	}
+	if !validImage {
+		return fmt.Errorf("imageid must be a valid ami available in the specified region")
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *AwsEc2Res) Init() error {
+	// create a client session for the AWS API
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	})
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating session")
+	}
+	obj.client = ec2.New(sess)
+
+	obj.wg = &sync.WaitGroup{}
```

Comment:
Since this is only used inside the Watch method, it doesn't need to be defined in the struct (or initialized here) and can be a local variable in watch, eg:  `wg := &sync.WaitGroup{}`